idx/memory/partitioned_idx.go

```diff
@@ -0,0 +1,536 @@
+package memory
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/grafana/metrictank/cluster"
+	"github.com/grafana/metrictank/idx"
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// Implements the the "MetricIndex" interface
+type PartitionedMemoryIdx struct {
+	Partition map[int32]*MemoryIdx
+}
+
+func NewPartitionedMemoryIdx() *PartitionedMemoryIdx {
+	idx := &PartitionedMemoryIdx{
+		Partition: make(map[int32]*MemoryIdx),
+	}
+	partitions := cluster.Manager.GetPartitions()
+	//log.Infof("PartitionedMemoryIdx: initializing with partitions: %v", partitions)
+	for _, p := range partitions {
+		idx.Partition[p] = NewMemoryIdx()
+	}
+	return idx
+}
+
+// Init initializes the index at startup and
+// blocks until the index is ready for use.
+func (p *PartitionedMemoryIdx) Init() error {
+	for _, m := range p.Partition {
+		//log.Infof("PartitionedMemoryIdx: initializing MemoryIdx for partition %d", part)
+		err := m.Init()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Stop shuts down the index.
+func (p *PartitionedMemoryIdx) Stop() {
+	for _, m := range p.Partition {
+		m.Stop()
+	}
+}
+
+// Update updates an existing archive, if found.
+// It returns whether it was found, and - if so - the (updated) existing archive and its old partition
+func (p *PartitionedMemoryIdx) Update(point schema.MetricPoint, partition int32) (idx.Archive, int32, bool) {
+	return p.Partition[partition].Update(point, partition)
+}
+
+// AddOrUpdate makes sure a metric is known in the index,
+// and should be called for every received metric.
+func (p *PartitionedMemoryIdx) AddOrUpdate(mkey schema.MKey, data *schema.MetricData, partition int32) (idx.Archive, int32, bool) {
+	return p.Partition[partition].AddOrUpdate(mkey, data, partition)
+}
+
+// UpdateArchive updates the archive information
+func (p *PartitionedMemoryIdx) UpdateArchive(archive idx.Archive) {
+	p.Partition[archive.Partition].UpdateArchive(archive)
+}
+
+// Get returns the archive for the requested id.
+func (p *PartitionedMemoryIdx) Get(key schema.MKey) (idx.Archive, bool) {
+	g, _ := errgroup.WithContext(context.Background())
+	resultChan := make(chan *idx.Archive, len(p.Partition))
+	for _, m := range p.Partition {
+		m := m
+		g.Go(func() error {
+			if a, ok := m.Get(key); ok {
+				resultChan <- &a
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		log.Errorf("memory-idx: failed to get Archive with key %v. %s", key, err)
+		return idx.Archive{}, false
+	}
+	close(resultChan)
+	result, ok := <-resultChan
+	if !ok {
+		return idx.Archive{}, false
+	}
+
+	return *result, true
+}
+
+// GetPath returns the archives under the given path.
+func (p *PartitionedMemoryIdx) GetPath(orgId uint32, path string) []idx.Archive {
+	g, _ := errgroup.WithContext(context.Background())
+	result := make([][]idx.Archive, len(p.Partition))
+	var i int
+	for _, m := range p.Partition {
```

Comment:
instead of `_` this could just be `i` and then the var wouldn't need to be initialized above nor increased further down, because its actual value seems to make no difference in this use case, only its uniqueness

Reply:
`p.Partition` is a map, not a slice.

Reply:
i know, so `i` will be the key of the map, but the resulting return value from the function should remain the same

Reply:
i is used as the index into the result slice.  You cant use the partition number for this.  Only the first shard would have partition numbers 0,1,2,3,4,5,6,7.  then next shard would have partitions 8,9,10,11,12,13,14,15

Reply:
I'm starting to feel like I must be missing something :joy:. Currently the key in `result` is an increasing int starting from `0`, where each value is unique for one partition. If you'd use the keys of `p.Partition` instead of that increasing int, then the keys in `result` wouldn't necessarily be beginning from `0` anymore because some MT consume the higher partitions (as you're describing). But my point is that the value of the keys in `result` is irrelevant because it is not used anywhere, so it doesn't matter if those keys are starting at `0` or some higher value. Honestly this change really isn't very important, it's only going to save the allocation of one int.

Reply:
@replay https://play.golang.org/p/Gu_xXmkNRuN